scripts/bumper/main.go

```diff
@@ -0,0 +1,320 @@
+				logger.WithError(err).Error("failed to write output")
+			}
+		}
+		if err := writer.Flush(); err != nil {
+			logger.WithError(err).Error("failed to flush output")
+		}
+	case synchronize:
+		for i, commit := range commits {
+			commitLogger := logger.WithField("commit", commit.Hash)
+			commitLogger.Infof("cherry-picking commit %d/%d", i+1, len(commits))
+			if err := cherryPick(ctx, commitLogger, commit); err != nil {
+				logger.WithError(err).Error("failed to cherry-pick commit")
+				break
+			}
+		}
+	}
+}
+
+type commit struct {
+	Date    time.Time `json:"date"`
+	Hash    string    `json:"hash,omitempty"`
+	Author  string    `json:"author,omitempty"`
+	Message string    `json:"message,omitempty"`
+	Repo    string    `json:"repo,omitempty"`
+}
+
+var repoRegex = regexp.MustCompile(`Upstream-repository: ([^ ]+)\n`)
+var commitRegex = regexp.MustCompile(`Upstream-commit: ([a-f0-9]+)\n`)
+
+func detectNewCommits(ctx context.Context, logger *logrus.Entry, stagingDir string) ([]commit, error) {
+	lastCommits := map[string]string{}
+	if err := fs.WalkDir(os.DirFS(stagingDir), ".", func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d == nil || !d.IsDir() {
+			return nil
+		}
+
+		if path == "." {
+			return nil
+		}
+		logger = logger.WithField("repo", path)
+		logger.Debug("detecting commits")
+		output, err := runCommand(logger, exec.CommandContext(ctx,
+			"git", "log",
+			"-n", "1",
+			"--grep", "Upstream-repository: "+path,
+			"--grep", "Upstream-commit",
+			"--all-match",
+			"--pretty=%B",
+			"--",
+			filepath.Join(stagingDir, path),
+		))
+		if err != nil {
+			return err
+		}
+		var lastCommit string
+		commitMatches := commitRegex.FindStringSubmatch(output)
+		if len(commitMatches) > 0 {
+			if len(commitMatches[0]) > 1 {
+				lastCommit = string(commitMatches[1])
+			}
+		}
+		if lastCommit != "" {
+			logger.WithField("commit", lastCommit).Debug("found last commit synchronized with staging")
+			lastCommits[path] = lastCommit
+		}
+
+		if path != "." {
+			return fs.SkipDir
+		}
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("failed to walk %s: %w", stagingDir, err)
+	}
+
+	var commits []commit
+	for repo, lastCommit := range lastCommits {
+		if _, err := runCommand(logger, exec.CommandContext(ctx,
+			"git", "fetch",
+			"git@github.com:operator-framework/"+repo,
+			"master",
+		)); err != nil {
+			return nil, err
+		}
+
+		output, err := runCommand(logger, exec.CommandContext(ctx,
+			"git", "log",
+			"--pretty=%H",
+			lastCommit+"...FETCH_HEAD",
+		))
+		if err != nil {
+			return nil, err
+		}
+
+		for _, line := range strings.Split(output, "\n") {
+			line = strings.TrimSpace(line)
+			if line != "" {
+				infoCmd := exec.CommandContext(ctx,
+					"git", "show",
+					line,
+					"--pretty=format:%H|%cI|%an|%s",
```

Comment:
A bit concerned here... mostly because I could see a `|` being used in the message, and the handling of multi-line messages.

Reply:
`go-get` might offer better interface to go, particularly the output.

Reply:
Changed to use a non-breaking space